### PR TITLE
Refactor engine controller to use structured logging

### PR DESCRIPTION
**Pull request by joshimoo**
By switching to structured logging and removing the common element
information from the individual log messages it becomes much simpler to
correlate a log statement with some code flow. The structured log fields
also allow us to easily filter logs based on specific element keys.

depends on longhorn/longhorn-manager#661
longhorn/longhorn#1724
longhorn/longhorn#1725

Signed-off-by: Joshua Moody <joshua.moody@rancher.com>

### CODE REVIEW

**Comment by joshimoo** [comment]
This depends on #661 and I will rebase this once #661 has been merged.

File: controller/engine_image_controller_test.go [hunk 1]
```diff
@@ -2,6 +2,7 @@ package controller
 
 import (
 	"fmt"
+	"github.com/sirupsen/logrus"
```
Comment (yasker) [inline controller/engine_image_controller_test.go:5]: Add a blank.

**Comment by joshimoo** [comment]
Please only review the last commit (engine-controller) the other code is covered by #661 (please add review comment there instead) and will not be part of this pr once #661 is committed

File: controller/engine_controller.go [hunk 2]
```diff
@@ -562,6 +564,7 @@ func (ec *EngineController) isMonitoring(e *longhorn.Engine) bool {
 func (ec *EngineController) startMonitoring(e *longhorn.Engine) {
 	stopCh := make(chan struct{})
 	monitor := &EngineMonitor{
+		logger:           ec.logger.WithField("engine", e.Name),
```
Comment (shuo-wu) [inline controller/engine_controller.go:567]: Would `engine-monitor` be better?

File: controller/engine_controller.go [hunk 3]
```diff
@@ -607,8 +610,8 @@ func (ec *EngineController) stopMonitoring(e *longhorn.Engine) {
 }
 
 func (m *EngineMonitor) Run() {
-	logrus.Debugf("Start monitoring %v", m.Name)
-	defer logrus.Debugf("Stop monitoring %v", m.Name)
+	m.logger.Debug("Start monitoring engine")
```
Comment (shuo-wu) [inline controller/engine_controller.go:613]: `Start monitoring`/`Stop monitoring` would be enough.

File: controller/engine_controller.go [hunk 4]
```diff
@@ -911,7 +914,7 @@ func syncWithRestoreStatus(engine *longhorn.Engine, rsMap map[string]*types.Rest
 
 	for _, status := range rsMap {
 		if status.Error != "" {
-			logrus.Warnf("Need to wait for the restore error handling for engine %v before the restore invocation: %v", engine.Name, status.Error)
+			log.Warnf("need to wait for the restore error handling for engine %v before the restore invocation: %v", engine.Name, status.Error)
```
Comment (shuo-wu) [inline controller/engine_controller.go:917]:
```
log.Warnf("need to wait for the restore error handling before the restore invocation: %v", status.Error)
```

File: controller/engine_controller.go [hunk 5]
```diff
@@ -1017,13 +1021,13 @@ func checkSizeBeforeRestoration(engine *longhorn.Engine, ds *datastore.DataStore
...
 			// The volume controller will update `engine.Spec.VolumeSize` later then trigger expansion call
-			logrus.Infof("engine monitor: Prepare to expand the size from %v to %v for DR volume %v", v.Spec.Size, bvSize, v.Name)
+			log.Infof("Prepare to expand the size from %v to %v for DR volume %v", v.Spec.Size, bvSize, v.Name)
 			v.Spec.Size = bvSize
 			if _, err := ds.UpdateVolume(v); err != nil {
 				if !datastore.ErrorIsConflict(err) {
 					return false, err
 				}
-				logrus.Debugf("engine monitor: Retrying update the volume %v size before restore", v.Name)
+				log.Debugf("Retrying size update for DR volume %v before restore", v.Name)
```
Comment (shuo-wu) [inline controller/engine_controller.go:1030]: Can we use `WithField("volume", e.Spec.VolumeName)`? Similar below.

**Review by shuo-wu — CHANGES_REQUESTED** [review]

File: controller/engine_controller.go (hunk 5 above)
Comment (joshimoo) [reply controller/engine_controller.go:1030]: yes, we can add it as a field.

File: controller/engine_controller.go (hunk 3 above)
Comment (joshimoo) [reply controller/engine_controller.go:613]: It's better to have `Start monitoring engine` where engine is defined as a field.
This way a search in the logs will be specific.

If you use a generic `start monitoring` in all the different components you will most likely end up with other components.
(ie instance-manager, backupstore, etc)

File: controller/engine_controller.go (hunk 2 above)
Comment (joshimoo) [reply controller/engine_controller.go:567]: The object we are concerned with is a `engine` not the monitor.

Comment (joshimoo) [reply controller/engine_controller.go:567]: Example, log print from below to make it more clear what it mean by that.
`log.Warnf("need to wait for the restore error handling before the restore invocation: %v", status.Error)`
this is concerned with a `engine` not a monitor, the monitor itself is only an internal construction that is not exposed to the outside.

Comment (shuo-wu) [reply controller/engine_controller.go:567]: Why is the `engine` preferred?

Comment (joshimoo) [reply controller/engine_controller.go:567]: See comment above?
We want to correlate actions across components that are involved with the `engine` object.

Maybe this helps?
It wouldn't make much sense calling it `engine-monitor` here then in the longhorn-engine/instance-manager project the field would be called `engine`?

Comment (joshimoo) [reply controller/engine_controller.go:567]: Let me flip the question, why do you think `engine-monitor` would be better?

Comment (shuo-wu) [reply controller/engine_controller.go:567]: Got it! Sorry for missing the comment, since the page was not refreshed when I replied to you.

Comment (shuo-wu) [reply controller/engine_controller.go:567]: I was thinking the original logger is based on the controller, hence this derived logger should base on the monitor.

**Comment by joshimoo** [comment]
addressed issues and rebased.

**Comment by joshimoo** [comment]
Rebased and Ready for review, added some more withError calls.

**Review by shuo-wu — APPROVED** [review]

**Pull request merged**